File: haskell/arith/smallStep.hs
-- Terms
data Term =
  TrueT |
  FalseT |
  IfT Term Term Term |
  ZeroT |
  SuccT Term |
  PredT Term |
  IsZeroT Term deriving Show

pretty :: Term -> String
pretty TrueT = "true"
pretty FalseT = "false"
pretty ZeroT = "0"
pretty (SuccT t1) | isNumericVal t1 = show $ 1 + read (pretty t1)
pretty t = "stuck \"" ++ show t ++ "\""

data Evaluation t = More t | Finished t

instance Functor Evaluation where
  fmap f (More t) = More (f t)
  fmap f (Finished t) = Finished (f t)

isNumericVal :: Term -> Bool
isNumericVal ZeroT = True
isNumericVal (SuccT t) = isNumericVal t
isNumericVal _ = False

evalStep :: Term -> Evaluation Term
evalStep t = case t of
  IfT TrueT t2 _ -> More t2
  IfT FalseT _ t3 -> More t3
  IfT t1 t2 t3 -> fmap (\t1' -> IfT t1' t2 t3) (evalStep t1)
  SuccT t1 -> fmap SuccT (evalStep t1)
  PredT ZeroT -> More ZeroT
  PredT (SuccT tt1) | isNumericVal tt1 -> More tt1
  PredT t1 -> fmap PredT (evalStep t1)
  IsZeroT ZeroT -> More TrueT
  IsZeroT (SuccT tt1) | isNumericVal tt1 -> More FalseT
  IsZeroT t1 | fmap IsZeroT (evalStep t1)
  _ -> Finished t

eval :: Term -> Term
eval t = case evalStep t of
  More t' -> eval t'
  Finished t' -> t'

testTerms = [
  IfT TrueT ZeroT (SuccT ZeroT),
  IfT FalseT ZeroT (SuccT ZeroT),
  IfT (IsZeroT ZeroT) (SuccT ZeroT) ZeroT,
  IsZeroT ZeroT,
  SuccT (PredT (SuccT ZeroT)),
  SuccT (IsZeroT (SuccT ZeroT))
  ]

main :: IO ()
main = mapM_ (
  \t -> do
    putStrLn $ show t ++ " -> " ++ (pretty $ eval t)
  ) testTerms

File: rust/arith/src/small_step.rs
use arith::*;

enum Evaluation {
  More(Term),
  Done(Term),
}

impl Evaluation {
  fn map<F: Fn(&Term) -> Term>(&self, f: F) -> Self {
    match self {
      Evaluation::More(t) => Evaluation::More(f(t)),
      Evaluation::Done(t) => Evaluation::Done(f(t)),
    }
  }
}

fn eval_step(term: &Term) -> Evaluation {
  match term {
    Term::If(box Term::True, t2, _t3) => Evaluation::More(*t2.clone()),
    Term::If(box Term::False, _t2, t3) => Evaluation::More(*t3.clone()),
    Term::If(t1, t2, t3) => {
      eval_step(t1).map(|t1_next| Term::If(box t1_next.clone(), t2.clone(), t3.clone()))
    }
    Term::Succ(t1) => eval_step(t1).map(|t1_next| Term::Succ(box t1_next.clone())),
    Term::Pred(box Term::Zero) => Evaluation::More(Term::Zero),
    Term::Pred(box Term::Succ(tt1)) if is_numeric_val(tt1) => Evaluation::More(*tt1.clone()),
    Term::Pred(t1) => eval_step(t1).map(|t1_next| Term::Pred(box t1_next.clone())),
    Term::IsZero(box Term::Zero) => Evaluation::More(Term::True),
    Term::IsZero(box Term::Succ(tt1)) if is_numeric_val(tt1) => Evaluation::More(Term::False),
    Term::IsZero(t1) => eval_step(t1).map(|t1_next| Term::IsZero(box t1_next.clone())),
    _ => Evaluation::Done(term.clone()),
  }
}

pub fn eval(term: &Term) -> Term {
  match eval_step(term) {
    Evaluation::More(t) => eval(&t),
    Evaluation::Done(t) => t,
  }
}
